### PR TITLE
[iot] Fixes on SAS token generation

### PR DESCRIPTION
**Pull request by ewertons**
Implemented URL-encoding of components of SAS tokens according to Azure IoT documentation:

For IoT Hub:
https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-security#security-token-structure

For Device Provisioning:
https://docs.microsoft.com/en-us/azure/iot-dps/concepts-symmetric-key-attestation#detailed-attestation-process

### CODE REVIEW

File: sdk/iot/common/inc/az_iot_common_internal.h [hunk 1]
```diff
@@ -32,6 +32,16 @@
  */
 AZ_NODISCARD int32_t _az_iot_u32toa_size(uint32_t number);
 
+/**
+ * @brief Copies the url-encoded content of `source` span into `destination`, returning the free remaining of `destination`.
+ *
+ * @param[in] destination The span where the url-encoded content is copied to.
+ * @param[in] source The span to url-encode and copy the content from.
+ * @param[out] out_remainder A slice of `destination` with the non-used portion of `destination`.
```
Comment (ericwol-msft) [inline sdk/iot/common/inc/az_iot_common_internal.h:40]: out_remainder A slice of `destination` with the non-used buffer portion of `destination`.

File: sdk/iot/provisioning/src/az_iot_provisioning_client_sas.c [hunk 2]
```diff
@@ -142,17 +128,24 @@ AZ_NODISCARD az_result az_iot_provisioning_client_sas_get_password(
   if (az_span_size(key_name) > 0)
   {
     // Key Name
+    AZ_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_password_span,
+      1 // AMPERSAND
```
Comment (danewalton) [inline sdk/iot/provisioning/src/az_iot_provisioning_client_sas.c:132]: nit: I prefer having each of these `1`'s as a call to `sizeof()` but not a blocker.

File: sdk/iot/common/inc/az_iot_common_internal.h [hunk 3]
```diff
@@ -32,6 +32,16 @@
  */
 AZ_NODISCARD int32_t _az_iot_u32toa_size(uint32_t number);
 
+/**
+ * @brief Copies the url-encoded content of `source` span into `destination`, returning the free remaining of `destination`.
+ *
+ * @param[in] destination The span where the url-encoded content is copied to.
+ * @param[in] source The span to url-encode and copy the content from.
```
Comment (ericwol-msft) [inline sdk/iot/common/inc/az_iot_common_internal.h:39]: @param[in] source The span to url-encode

_you use "copy" too much_

File: sdk/iot/common/inc/az_iot_common_internal.h [hunk 4]
```diff
@@ -32,6 +32,16 @@
  */
 AZ_NODISCARD int32_t _az_iot_u32toa_size(uint32_t number);
 
+/**
+ * @brief Copies the url-encoded content of `source` span into `destination`, returning the free remaining of `destination`.
+ *
+ * @param[in] destination The span where the url-encoded content is copied to.
```
Comment (ericwol-msft) [inline sdk/iot/common/inc/az_iot_common_internal.h:38]: @param[in] destination The span where the source is url-encoded to.

File: sdk/iot/provisioning/src/az_iot_provisioning_client_sas.c [hunk 5]
```diff
@@ -46,24 +46,22 @@ AZ_NODISCARD az_result az_iot_provisioning_client_sas_get_signature(
...
 
-  az_span remainder = signature;
-  remainder = az_span_copy(remainder, client->_internal.id_scope);
+  AZ_RETURN_IF_NOT_ENOUGH_SIZE(signature, az_span_size(resources_string));
   remainder = az_span_copy(remainder, resources_string);
-  remainder = az_span_copy(remainder, client->_internal.registration_id);
+
+  AZ_RETURN_IF_FAILED(_az_span_copy_url_encode(remainder, client->_internal.registration_id, &remainder));
+  
+  AZ_RETURN_IF_NOT_ENOUGH_SIZE(signature, 1 /* LF */);
```
Comment (danewalton) [inline sdk/iot/provisioning/src/az_iot_provisioning_client_sas.c:59]: ?
```suggestion
  AZ_RETURN_IF_NOT_ENOUGH_SIZE(remainder, 1 /* LF */);
```

File: sdk/iot/provisioning/src/az_iot_provisioning_client_sas.c [hunk 6]
```diff
@@ -46,24 +46,22 @@ AZ_NODISCARD az_result az_iot_provisioning_client_sas_get_signature(
...
-      + _az_iot_u32toa_size(token_expiration_epoch_time);
+  az_span remainder = signature;
+  int32_t signature_size = az_span_size(signature);
 
-  AZ_RETURN_IF_NOT_ENOUGH_SIZE(signature, required_size);
+  AZ_RETURN_IF_FAILED(_az_span_copy_url_encode(remainder, client->_internal.id_scope, &remainder));
 
-  az_span remainder = signature;
-  remainder = az_span_copy(remainder, client->_internal.id_scope);
+  AZ_RETURN_IF_NOT_ENOUGH_SIZE(signature, az_span_size(resources_string));
```
Comment (danewalton) [inline sdk/iot/provisioning/src/az_iot_provisioning_client_sas.c:54]: ?
```suggestion
  AZ_RETURN_IF_NOT_ENOUGH_SIZE(remainder, az_span_size(resources_string));
```

File: sdk/iot/common/inc/az_iot_common_internal.h [hunk 7]
```diff
@@ -32,6 +32,16 @@
...
+ * @param[in] source The span to url-encode and copy the content from.
+ * @param[out] out_remainder A slice of `destination` with the non-used portion of `destination`.
+ * @return An `az_result` value.
+ */
+AZ_NODISCARD az_result _az_span_copy_url_encode(az_span destination, az_span source, az_span* out_remainder);
```
Comment (ericwol-msft) [inline sdk/iot/common/inc/az_iot_common_internal.h:43]: _az_iot_span_url_encode ?

File: sdk/iot/provisioning/src/az_iot_provisioning_client_sas.c [hunk 8]
```diff
@@ -142,17 +128,24 @@ AZ_NODISCARD az_result az_iot_provisioning_client_sas_get_password(
...
     mqtt_password_span = az_span_copy(mqtt_password_span, key_name);
   }
 
+  AZ_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_password_span, 1 /* NULL TERMINATOR */);
   mqtt_password_span = az_span_copy_u8(mqtt_password_span, STRING_NULL_TERMINATOR);
 
   if (out_mqtt_password_length != NULL)
   {
-    *out_mqtt_password_length = ((size_t)required_size - 1);
+    *out_mqtt_password_length = ((size_t)mqtt_password_size - (size_t)az_span_size(mqtt_password_span) - 1);
```
Comment (danewalton) [inline sdk/iot/provisioning/src/az_iot_provisioning_client_sas.c:148]: comment for `NULL` terminator / call to `sizeof()` for subtraction.

File: sdk/iot/common/inc/az_iot_common_internal.h (hunk 1 above)
Comment (ewertons) [reply sdk/iot/common/inc/az_iot_common_internal.h:40]: Taken.

File: sdk/iot/common/inc/az_iot_common_internal.h (hunk 3 above)
Comment (ewertons) [reply sdk/iot/common/inc/az_iot_common_internal.h:39]: Sorry.

File: sdk/iot/common/inc/az_iot_common_internal.h (hunk 4 above)
Comment (ewertons) [reply sdk/iot/common/inc/az_iot_common_internal.h:38]: Fixed.

File: sdk/iot/provisioning/src/az_iot_provisioning_client_sas.c (hunk 2 above)
Comment (ewertons) [reply sdk/iot/provisioning/src/az_iot_provisioning_client_sas.c:132]: I tried using sizeof but had some issues. Will address that in separate.

File: sdk/iot/provisioning/src/az_iot_provisioning_client_sas.c (hunk 5 above)
Comment (ewertons) [reply sdk/iot/provisioning/src/az_iot_provisioning_client_sas.c:59]: Goodness, that's a big miss. Thanks for catching it.

File: sdk/iot/provisioning/src/az_iot_provisioning_client_sas.c (hunk 6 above)
Comment (ewertons) [reply sdk/iot/provisioning/src/az_iot_provisioning_client_sas.c:54]: Done.

File: sdk/iot/provisioning/src/az_iot_provisioning_client_sas.c (hunk 8 above)
Comment (ewertons) [reply sdk/iot/provisioning/src/az_iot_provisioning_client_sas.c:148]: Added comment.

File: sdk/iot/common/inc/az_iot_common_internal.h (hunk 7 above)
Comment (ewertons) [reply sdk/iot/common/inc/az_iot_common_internal.h:43]: This one is different. It url encodes, copies and **slices** the result to give the remaining buffer.

**Pull request merged**